src/main/java/dev/aziz/grocerystore/handlers/PriceHandler.java

```diff
@@ -1,28 +1,91 @@
 import java.util.List;
 
+@Slf4j
 public class PriceHandler {
 
-    private List<BasketItem> basketItems;
+    private final List<UserPromotion> userPromotions;
 
-    public PriceHandler(List<BasketItem> basketItems) {
+    private final List<PromotionConfig> promotionConfigs;
+
+    private final List<BasketItem> basketItems;
+
+    public PriceHandler(List<UserPromotion> userPromotions,
+                        List<PromotionConfig> promotionConfigs,
+                        List<BasketItem> basketItems) {
+        this.userPromotions = userPromotions;
+        this.promotionConfigs = promotionConfigs;
         this.basketItems = basketItems;
     }
 
     public BigDecimal computePrice() {
-        BigDecimal totalPrice = new BigDecimal(BigInteger.ZERO);
-        for (BasketItem basketItem : basketItems) {
-            totalPrice = totalPrice.add(
-                    basketItem.getItem().getPrice().multiply(BigDecimal.valueOf(basketItem.getAmount()))
-            );
-        }
+        BigDecimal totalPrice = BigDecimal.ZERO;
+
+        for (BasketItem item : basketItems) {
+            if (hasItemPromotion(item)) {
+                log.info("hasItemPromotion(item) condition passed for this item: {}.", item.getItem().getName());
 
+                PromotionConfig itemPromotion = null;
+
+                for (UserPromotion promotion : userPromotions) {
+                    if (item.getId() == promotion.getPromotionConfig().getItem().getId()) {
+                        itemPromotion = promotion.getPromotionConfig();
+                    }
+                }
+                if (itemPromotion != null) {
+                    totalPrice = totalPrice.add(computePriceWithPromotion(item, itemPromotion));
+                } else {
+                    BigDecimal add = item.getItem().getPrice().multiply(BigDecimal.valueOf(item.getAmount()));
+                    log.info("Total items without promotion: {}, and total price: {}", item.getAmount(), add);
+                    totalPrice = totalPrice.add(add);
+                }
+            } else {
+                BigDecimal add = item.getItem().getPrice().multiply(BigDecimal.valueOf(item.getAmount()));
+                log.info("Total items without promotion: {}, and total price: {}", item.getAmount(), add);
+                totalPrice = totalPrice.add(add);
+            }
+        }
         return totalPrice;
     }
 
+    private BigDecimal computePriceWithPromotion(BasketItem basketItem, PromotionConfig promotionConfig) {
+        Integer minAmount = promotionConfig.getMinimumAmount();
+        Integer freeAmount = promotionConfig.getFreeAmount();
+        Instant date = promotionConfig.getEndDate();
+        int value = 0;
+
+        Instant now = Instant.now();
+        value = now.compareTo(date);
+        int free = 0;
+
+        if (value < 0) {
+            if (basketItem.getAmount() >= minAmount) {
+                free = (basketItem.getAmount() / minAmount) * freeAmount;
```

Comment:
I don't think this will work. If I have a minimum amount of 2 items for 1 free item, you need 3 items in the basket.

This will result into 1.5 item as free.

Just make:
```
free = 0;
if (basket.amount > minAmount) {
  free = freeAmount;
}
paidItems = basket.amount - free;
```